src/bin/pgcopydb/ld_apply.c

```diff
@@ -1125,15 +1163,41 @@ setupReplicationOrigin(StreamApplyContext *context, bool logSQL)
 		return false;
 	}
 
-	if (!pgsql_replication_origin_progress(pgsql,
-										   nodeName,
-										   true,
-										   &(context->previousLSN)))
+	/*
+	 * Fetch the replication origin LSN tracking, which is maintained in a
+	 * transactional fashion with the SQL that's been replayed. It's the
+	 * authoritative value for progress at reconnect, given that we use
+	 * synchronous_commit off.
+	 */
+	uint64_t originLSN = InvalidXLogRecPtr;
+
+	if (!pgsql_replication_origin_progress(pgsql, nodeName, true, &originLSN))
 	{
 		/* errors have already been logged */
 		return false;
 	}
 
+	if (context->previousLSN == InvalidXLogRecPtr)
+	{
+		log_info("Setting up previous LSN from "
+				 "replication origin \"%s\" progress at %X/%X",
+				 nodeName,
+				 LSN_FORMAT_ARGS(originLSN));
+
+		context->previousLSN = originLSN;
+	}
+	else if (context->previousLSN != originLSN)
```

Comment:
I'm wondering when this condition would hit? Wouldn't that `context->previousLSN` always `InvalidXLogRecPtr` ?

Reply:
It could have been initialized from reading the value in the sentinel at startup, when re-launching the pgcopydb command from a previous run. I have added a comment explaining that, thanks.